### PR TITLE
Virtual Deserialization v0.1.3

### PR DESCRIPTION
**Pull request by jtenner**
We can now deserialize a reference as a base class without hitting runtime assertion errors.

### CODE REVIEW

File: packages/assembly/index.ts [hunk 1]
```diff
@@ -850,7 +850,7 @@ export namespace ASON {
       // all the references have been allocated, let's get entry 0 and validate type info
       let entry0 = changetype<usize>(entryMap.get(0));
       if (isReference<T>()) {
-        assert(getObjectType(entry0) == idof<T>());
+        assert(getObjectType(entry0) == idof<T>() || __instanceof(entry0, idof<T>()));
```
Comment (jtenner) [inline packages/assembly/index.ts:853]: This is the line in question. `__instanceof` ships with the `assemblyscript` runtime as a means to check `a instanceof B` when `B` is a base class.

File: packages/test/assembly/__tests__/index.spec.ts [hunk 2]
```diff
@@ -411,3 +417,13 @@ function testCallbacks(): void {
   let b = ASON.deserialize<() => void>(buffer);
   b();
 }
+
+function testVirtual(): void {
+  let child = new Child();
+  child.a = 42;
+  let buffer = ASON.serialize(child);
+  let result = ASON.deserialize<Base>(buffer);
```
Comment (jtenner) [inline packages/test/assembly/__tests__/index.spec.ts:425]: This is the behavior that would have previously caused an error, however, this test now passes.

**Comment by jtenner** [comment]
CI doesn't pass because of a version problem. I need to fix CI.

**Comment by jtenner** [comment]
```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: ason@0.0.4
npm ERR! Found: assemblyscript@0.19.22
npm ERR! node_modules/assemblyscript
npm ERR!   dev assemblyscript@"^0.19.22" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer assemblyscript@"^0.18.31" from visitor-as@0.6.0
npm ERR! node_modules/visitor-as
npm ERR!   visitor-as@"^0.6.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /home/runner/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2022-01-08T20_54_47_356Z-debug.log
Error: Process completed with exit code 1.
```

**Review by RedDwarfian — APPROVED** [review]
Good that `__instanceof` exists. I would hate to have had to roll our own version of that.

**Pull request merged**